### PR TITLE
backend: run tests as a stock user

### PR DESCRIPTION
**Pull request by guewen**
We have to be sure that all scenarii can be run with a normal
stock user, without needing manager permissions.

Now the test are all run as normal stock user, to verify this.

The consequences on writing tests:

* Records created or written in a test setup must use sudo()
  if the user has no permission on these models.
* Tests setUps should no longer extend setUpClass but setUpClassVars
  and setUpClassBaseData, which already have an environment with
  the stock user.
* Be wary of creating records before setUpClassUsers is called, because
  it their "env.user" would be admin and could lead to inconsistencies
  in tests.

The services / actions which needed a correction for stock users:

* Creating an inventory (cluster picking stock issue, ...)
* Validating an inventory (cluster picking stock issue, ...)
* In cluster's picking when a package is changed,
  'StockQuant.inventory_quantity' is updated, this field automatically
  create moves to adjust quantities. However, it was ignored if the user
  doesn't have the 'stock manager' group.
* OpenAPI: the swagger UI could not work without being admin because it
  reads the demo auth API key, which is obviously restricted. Reading it
  as sudo is safe as it's a demo key.

### CODE REVIEW

File: shopfloor/models/stock_quant.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+    def _is_inventory_mode(self):
+        """ Used to control whether a quant was written on or created during an
+        "inventory session", meaning a mode where we need to create the stock.move
+        record necessary to be consistent with the `inventory_quantity` field.
+        """
+        # The default method check if we have the stock.group_stock_manager
+        # group, however, we want to force using this mode from shopfloor
+        # (cluster picking) when sudo is used and the user is a stock user.
+        if self.env.context.get("inventory_mode") is True and self.env.su:
```
Comment (sebalix) [inline shopfloor/models/stock_quant.py:15]: Is `is True` mandatory here?

File: shopfloor/tests/test_checkout_select.py [hunk 2]
```diff
@@ -52,7 +52,7 @@ def _test_error(self, picking, msg):
 
     def test_select_error_not_found(self):
         picking = self._create_picking()
-        picking.unlink()
+        picking.sudo().unlink()
         self._test_error(picking, "This transfer does not exist anymore.")
```
Comment (sebalix) [inline shopfloor/tests/test_checkout_select.py:56]: I'm not sure to understand how the `_test_error` is working as expected when it performs the call `_picking_summary_data` which "jsonify" the picking, I was thinking it should crash if the recordset data are read while the it is actually removed.
But if it works, I won't look further... :)

**Review by sebalix — APPROVED** [review]
Perhaps part of your commit message / PR description have to be written in `test/common.py` for the future, to know how to write new tests? Pre-approving as it seems OK for the little I know about this module.

File: shopfloor/models/stock_quant.py (hunk 1 above)
Comment (guewen) [reply shopfloor/models/stock_quant.py:15]: I don't think it should be, but I copied from the base method: https://github.com/odoo/odoo/blob/90008eaee70744a6080ea2e7359a0bf0a6eb7714/addons/stock/models/stock_quant.py#L542-L547

So I prefer to have the exact same condition.

File: shopfloor/tests/test_checkout_select.py (hunk 2 above)
Comment (guewen) [reply shopfloor/tests/test_checkout_select.py:56]: It's actually what it tests: `_picking_summary_data` should not return the jsonified picking but return to selection of picking (state: `manual_selection`) with an error message ("This transfer does not exist anymore.") and the details of pickings that still exist (`self.picking` here, which we didn't delete)

**Comment by guewen** [comment]
> Perhaps part of your commit message / PR description have to be written in `test/common.py` for the future, to know how to write new tests

I'll do this. If this PR blocks you to start working on delivery, you can merge already and I can do it in a follow-up PR... ;)

**Review by simahawk — APPROVED** [review]

**Comment by simahawk** [comment]
let's merge pls :)
We can adjust small things later :kissing:

**Pull request merged**